jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluator.java

```diff
@@ -74,16 +78,17 @@ public Object evaluate(ProcessContext context) throws Exception {
             VariableScopeInstance variableScope = (VariableScopeInstance) ((WorkflowProcessInstance)context.getProcessInstance())
                     .getContextInstance(VariableScope.VARIABLE_SCOPE);
     
-            if (variables != null ) {
-                variables.putAll(variableScope.getVariables());
-            }
+            variables.putAll(variableScope.getVariables());
         }
         DMNRuntime runtime = ((KieSession) context.getKieRuntime()).getKieRuntime(DMNRuntime.class);
         List<FEELProfile> profiles = (List)((DMNRuntimeImpl) runtime).getProfiles();
         FEEL feel = FEEL.newInstance(runtime.getRootClassLoader(), profiles);
+        FeelReturnValueEvaluatorListener listener = new FeelReturnValueEvaluatorListener();
+        feel.addListener(listener);
         
         Object value = feel.evaluate(expr, variables);
 
+        listener.getEvents().forEach(FeelReturnValueEvaluator::processEvents);
```

Comment:
If we want to process all events generated by the listener and do not fail with the first error, I guess we will need something like
StringBuilder errorMessage = new StringBuilder();
for (FeelEvent : listerner.getEvents()) {
    processEvent(event,errorMessage);
}
if (errorMessage.length()>0) { throw new FeelReturnValueEvaluatorException(errorMessage);

and 
private static void process(FEELEvent event, StringBuilder messageBuilder) {

....
 case ERROR:
                StringBuilder messageBuilder = new StringBuilder(event.getSeverity().toString()).append(" ").append(event.getMessage());
                if (event.getOffendingSymbol() != null) {
                    messageBuilder.append(" ( offending symbol: '").append(event.getOffendingSymbol()).append("' )");
                }
                if (event.getSourceException() != null) {
                    messageBuilder.append("  ").append(event.getSourceException().getMessage());
                }
                 LOG.error("Event: {} Error: {}", event, messageBuilder);
.....
}                  
    


}
     
}

Reply:
If we want to fail with the first error, then I will change the listener to generate the exception and add a getException to it, so there is  not need to have a list inside the listener and the code looks like
 Object value = feel.evaluate(expr, variables);
if (listener.getException() != null) 
   throw listener.getException();

And listener code 

public class FeelReturnValueEvaluatorListener implements FEELEventListener {

    private FeelReturnValueEvaluatorException ;

    @Override
    public void onEvent(FEELEvent event) {
         switch (event.getSeverity()) {
            case ERROR:
                LOG.error("{}", event);
                StringBuilder messageBuilder = new StringBuilder(event.getSeverity().toString()).append(" ").append(event.getMessage());
                if (event.getOffendingSymbol() != null) {
                    messageBuilder.append(" ( offending symbol: '").append(event.getOffendingSymbol()).append("' )");
                }
                if (event.getSourceException() != null) {
                    messageBuilder.append("  ").append(event.getSourceException().getMessage());
                  exception = FeelReturnValueEvaluatorException(messageBuilder.toString(), event.getSourceException()); 
                   break;
                }
               else {
                 exception =  new FeelReturnValueEvaluatorException(messageBuilder.toString());
                }
            case TRACE:
                LOG.debug("{}", event);
                break;
            case WARN:
                LOG.warn("{}", event);
                break;
            case INFO:
            default:
                LOG.info("{}", event);
               break;
    }

    public  getException() {
        return exception;
    }
}